copybookreader-plugins/src/main/java/co/cask/hydrator/plugin/batch/CopybookSource.java

```diff
@@ -0,0 +1,251 @@
+    // Set the input file path for the job
+    CopybookInputFormat.setInputPaths(job, config.binaryFilePath);
+    CopybookInputFormat.setMaxInputSplitSize(job, config.maxSplitSize);
+    context.setInput(Input.of(config.referenceName, new SourceInputFormatProvider(CopybookInputFormat.class,
+                                                                                  job.getConfiguration())));
+  }
+
+  @Override
+  public void transform(KeyValue<LongWritable, Map<String, AbstractFieldValue>> input,
+                        Emitter<StructuredRecord> emitter)
+    throws Exception {
+    Map<String, AbstractFieldValue> value = input.getValue();
+    if (outputSchema == null) {
+      outputSchema = getSchema(value.keySet());
+    }
+    StructuredRecord.Builder builder = StructuredRecord.builder(outputSchema);
+    for (Schema.Field field : outputSchema.getFields()) {
+      String fieldName = field.getName();
+      if (value.containsKey(fieldName)) {
+        builder.set(fieldName, parseValue(field, value.get(fieldName)));
+      }
+    }
+    emitter.emit(builder.build());
+  }
+
+  /**
+   * Convert each field to data type as specified by user in the output schema
+   *
+   * @param field Schema.Field object defining the field name and properties
+   * @param value value of field to be converted to datatype specified by user
+   * @return value casted to the required data type
+   * @throws IOException
+   */
+  private Object parseValue(Schema.Field field, @Nullable AbstractFieldValue value) throws IOException {
+    String fieldName = field.getName();
+    Schema fieldSchema = field.getSchema();
+    if (value == null) {
+      if (!fieldSchema.isNullable()) {
+        throw new IllegalArgumentException("NULL value found for non-nullable field : " + fieldName);
+      }
+      return null;
+    }
+    Schema.Type fieldType = fieldSchema.isNullable() ? fieldSchema.getNonNullable().getType() : fieldSchema.getType();
+    switch (fieldType) {
+      case NULL:
+        return null;
+      case INT:
+        return value.asInt();
+      case DOUBLE:
+        return value.asDouble();
+      case BOOLEAN:
+        return value.asBoolean();
+      case LONG:
+        return value.asLong();
+      case FLOAT:
+        return value.asFloat();
+      case STRING:
+        return value.asString();
+      default:
+        throw new IOException(String.format("Unsupported schema: %s for field: \'%s\'", field.getSchema(),
+                                            field.getName()));
+    }
+  }
+
+  /**
+   * Create schema from list field names retrieved from CopybookFiles
+   *
+   * @param fieldsSet set of field names extracted from Copybook
+   * @return output schema fields
+   */
+  private Schema getSchema(Set<String> fieldsSet) {
+    List<Schema.Field> fields = Lists.newArrayList();
+    for (String field : fieldsSet) {
+      fields.add(Schema.Field.of(field, Schema.nullableOf(Schema.of(Schema.Type.STRING))));
```

Comment:
this doesn't seem correct, output schema has everything as a nullable string, but parseValue() returns all sorts of different non-string types.